tpu-client/src/tpu_client.rs

```diff
@@ -95,6 +106,70 @@ where
         self.invoke(self.tpu_client.try_send_transaction(transaction))
     }
 
+    /// Serialize and send transaction to the current and upcoming leader TPUs according to fanout
+    /// size
+    /// Attempt to send and confirm tx "attempts" times
+    /// Wait for signature confirmation before returning
+    /// Return the transaction signature
+    /// NOTE: send_wire_transaction() and try_send_transaction() above both fail in a specific case when used in LocalCluster
+    /// They both invoke the nonblocking TPUClient and both fail when calling "transfer_with_client()" multiple times
+    /// I do not full understand WHY the nonblocking TPUClient fails in this specific case. But the method defined below
+    /// does work although it has only been tested in LocalCluster integration tests
+    pub fn send_and_confirm_transaction_with_retries<T: Signers + ?Sized>(
```

Comment:
What is the minimal amount of logic in this function that is necessary to avoid the `ConnectError(EndpointStopping)` error in local cluster?
Is it getting the ConnectionCache and using `send_data_async()` directly that resolves the error?

I am wondering if we can put the retry/resign/confirm logic in a local-cluster function, instead of adding such a new, heavy api in `TpuClient`.

Reply:
good question/point. I'll look into this. I had designed this to be similar to what `ThinClient` had with its retries. But I guess there is a reason we are getting rid of `ThinClient` lol.

Reply:
update: we do need some retry logic here. using the `ConnectionCache` and `send_data_async()` alone does not resolve the error. Going to try and put retry logic in a separate LocalCluster function